Ability to load fixture state for e2e tests

whymarrh opened the pull request:
Closes #6154

This PR adds the ability for us to write e2e tests that use fixture data for both the extension's background state and the current state of the blockchain (via [Ganache][1]). Fortunately we compute the initial Redux store contents (the front-end state) from the initial background state so loading the background state from a fixture is currently synonymous with loading the Redux state.

### Overview

As outlined in #6154:

> We need to create infrastructure and tooling that allows us to run our e2e tests from an existing state.

> We need a solution for the background that allows us to dynamically load state during test runs.

> Once this task is complete, our e2e tests will be able to start the app at a state given in [JSON] file (plus [Ganache]) and run tests from there.

This PR introduces a way to write and organize fixtures into a set of directories that contain both the background and Ganache state as well as a `load` function that takes a path to a fixture directory, loads that state and runs a set of assertions in that context.

By example:

```js
describe('MetaMask Browser Extension from existing state', function () {
  it('displays the lock screen', async () => {
    await load('onboarded', async ({driver}) => {
      const button = await driver.findElement('.unlock-page__title')
      await button.click()
    })
  })
})
```

This approach allows us to start the app in a given state and make assertions from there. The implementation (detailed below) allows us to execute each of the tests in parallel. The arguments passed to the callback currently include a wrapper around a `WebDriver` instance that exposes helper functions similar to those we've already been using (e.g. functions handling timeouts and retries). The helper functions on the `WebDriver` wrapper can be expanded to include any sugar/convenience methods we'd like (see also #6155, #6156).

### The implementation

The background loads its state using the `LocalStore` wrapper for the `storage.local` API:

https://github.com/MetaMask/metamask-extension/blob/160f1961654632654072759f0aa116b43a97f35b/app/scripts/background.js#L182-L186

#### Background

The e2e tests now spin up a local HTTP server that provides fixture state to clients at pre-determined URLs. The extension now has an e2e flag set at build time that is used to set the application down e2e-specific flows which we're using in the background to load the initial state instead from the network (from said server). This setup keeps everything local but does introduce network latency where there wasn't any before only during e2e tests (a small cost). When built without the flag set the application behaves no differently.

Each time the `load` function is called the fixture server is pointed at a new fixture directory. This requires that the _startup_ of each test be sequential but allows the tests to _run_ in parallel after the initial state is loaded (i.e. all of the assertions, the majority of the test).

#### Ganache

Instead of using ganache-cli from the command line, the e2e tests now start ganache's server during each `load` invocation. At the same time the tests spin up a local server for the fixture data, we also start a ganache-cli server instance with the options outlined in the `${fixturesDirectory}/db.js` file (note that this is a JS file instead of JSON to be able to things like `path.resolve`). Once the server is up and running, we can interact with the RPC endpoint. Because the server is started on any available port, a `ganache` reference is passed through to the tests to allow them to setup the RPC endpoint as needed.

### The state of this PR (tasks)

The following task list will be updated as this PR is updated:

- Ability to load background state in
    - [x] Chrome
    - [x] Firefox
- [x] Ability to load Ganache state

### Testing this PR (no pun intended)

This PR includes an example spec file (`e2e/tests/example.spec.js`) which outlines the API future tests can be written with. (It might be deleted before this is merged?) To run the example spec file:

Firefox (the default browser):

```
npx mocha --no-timeouts e2e/tests/example.spec.js
```

Firefox, explicitly:

```
BROWSER=firefox npx mocha --no-timeouts e2e/tests/example.spec.js
```

Chrome:

```
BROWSER=chrome npx mocha --no-timeouts e2e/tests/example.spec.js
```

---
### Alternative approaches to pre-loading background state

Briefly on the alternative approaches we've considered for this functionality:

#### Editing the extension source files before loading it

([A recap of a discussion in Slack.][2])

The old e2e tests did mutate background state by prepending the background file with a global variable assignment, setting the `METAMASK_TEST_CONFIG`:

https://github.com/MetaMask/metamask-extension/blob/cb2698d20eae273d372d03e11fa765a91c330c17/test/e2e/func.js#L28-L37

This approach doesn't support the ability for us to run tests in parallel. We could update it to include writing out a separate build of the extension to disk _N_ times for _N_ tests and edit each differently but that would be a lot of I/O during what is already an already slow test suite.

#### Preloading the extension state before loading the extension

The ideal approach to loading background would be to have the browser know about the state before the extension is installed. This is easily possible with Firefox but less easy to do with Chrome. Firefox stores extension state in a single text file on disk whereas Chrome stores its extension state across multiple binary [leveldb](http://leveldb.org) databases.<sup>[\[3\]][3]</sup>

  [1]:https://truffleframework.com/docs/ganache/overview
  [2]:https://consensys.slack.com/archives/G1L7H42BT/p1551233197042000?thread_ts=1550786382.026900&cid=G1L7H42BT
  [3]:https://consensys.slack.com/archives/G1L7H42BT/p1551229029040600?thread_ts=1550786382.026900&cid=G1L7H42BT

whymarrh committed:
Add a ReadOnlyNetworkStore for pulling state from the network

kumavis commented:
We have an API available for testing/tooling in the ui set the locale, could use it to override the state.

If we give it more powerful features like that we should consider deleting the api object after some time out, so it can only be used at boot and not abused at runtime

danjm commented on e2e/webdriver/chrome.js:
```diff
@@ -0,0 +1,60 @@
+      .addArguments([
+        `load-extension=${extensionPath}`,
+      ])
+      .setChromeBinaryPath('/Applications/Google Chrome Canary.app/Contents/MacOS/Google Chrome Canary')
```
This is an OS, and possibly individual system, dependent path

whymarrh replied:
My bad, good catch

whymarrh replied:
Fixed

whymarrh committed:
Load background state fixtures in e2e tests

whymarrh commented:
This PR currently supports the ability to load background state in **both** Firefox and Chrome

whymarrh commented on app/scripts/background.js:
```diff
@@ -65,7 +65,7 @@ const openMetamaskTabsIDs = {}
 // state persistence
 const diskStore = new LocalStorageStore({ storageKey: STORAGE_KEY })
-const localStore = new LocalStore()
+const localStore = new ReadOnlyNetworkStore()
```
This needs to be conditional on some environment variable

whymarrh replied:
But I think this can be merged in to the destination branch (`e2e-2.0`) for now

danjm replied:
> But I think this can be merged in to the destination branch (e2e-2.0) for now

We should create an issue or some sort of hard reminder to fix this before merging the destination branch to `develop`

whymarrh replied:
Created https://github.com/MetaMask/metamask-extension/issues/6324

whymarrh committed:
Load Ganache state with options during e2e tests

whymarrh commented:
77eb21a adds the ability to start Ganache from options (which includes an existing db path) during e2e tests. The Ganache wrapper (i.e., helper functions) hasn't been fleshed out yet since I don't have a good sense of what its API should be—that'll be something that comes out of #6157 as well. What we can do is interact with Ganache's RPC endpoint on some port.

whymarrh committed:
Reject Ganache#quit if the server is not running

whymarrh committed:
Centre the notification in the current window (#6307)

wmsouza committed:
Fixes popups not showing when screen size is odd (#6312)

tmashuang committed:
Fix typos in English messages (#6317)

danjm commented on e2e/index.js:
```diff
@@ -0,0 +1,40 @@
+const rc = new ReferenceCounter(async () => fixtureServer.start(), async () => fixtureServer.stop())
+
+const load = async function load (fixturesDirectory, callback) {
+  await rc.tick(async () => {
```
Why do we need the reference counter? Is it to handle multiple simultaneous calls of the fixture server, such that only one instance of the server can ever be running at a time?

whymarrh replied:
Yup. We can only have (and only need) the single server on port `12345` so we want to make sure it's only started once and that it's shutdown when the tests end. I'm open to better ways of doing this.

danjm commented on e2e/index.js:
```diff
@@ -0,0 +1,40 @@
+}
+
+module.exports = {
+  load,
```
Can we give `load` a slightly more unique name so that it is easy to grep for its uses in the codebase?

whymarrh replied:
That's... a really good suggestion. I'll do `withFixtures`?

whymarrh commented on app/scripts/lib/network-store.js:
```diff
@@ -0,0 +1,39 @@
+   * @return {Promise<void>}
+   */
+  async set (state) {
+    // Pass
```
Because it's ephemeral—during the lifetime of the app we don't need to bother saving the state that's set here

whymarrh commented on ui/app/store.js:
```diff
@@ -9,7 +9,7 @@ global.METAMASK_DEBUG = process.env.METAMASK_DEBUG
 module.exports = configureStore
 
 const loggerMiddleware = createLogger({
-  predicate: () => global.METAMASK_DEBUG,
```
Yup, removed. Good catch, this was for testing.

whymarrh commented on ui/index.js:
```diff
@@ -9,7 +9,7 @@ const log = require('loglevel')
 
 module.exports = launchMetamaskUi
 
-log.setLevel(global.METAMASK_DEBUG ? 'debug' : 'warn')
```
Yup, removed. Good catch, this was for testing.

whymarrh commented on app/scripts/lib/network-store.js:
```diff
@@ -0,0 +1,39 @@
+   */
+  async get () {
+    try {
+      const response = await fetch('http://localhost:12345/state.json')
```
I'll extract this into a constant, good point. I think I'll leave this hard-coded though as this is something that we need to share and keep the same between the app (which could be pre-built into `dist/`) and the e2e logic. It's _possible_ to have it dynamically set but I feel like this is easy enough to start with and should we _need_ to we can make it dynamic later.

whymarrh commented on e2e/fixtures/index.js:
```diff
@@ -0,0 +1,79 @@
+  async start () {
+    return new Promise((resolve) => {
+      const options = {
+        host: 'localhost',
```
Constants for these as well

whymarrh committed:
Rename load as withFixtures for e2e test fixtures

whymarrh committed:
Extract fixture server host and port into constants

The constants are mirrored between the e2e logic (`e2e/`) and the app logic
in the ReadOnlyNetworkStore (`app/scripts/lib/network-store.js`) as they need
to be the same in both places. These could eventually be made dynamic but the
current build process isn't kind to that idea (in that the app is compiled
and run separately from the e2e tests, which aren't compiled at all).

danfinlay committed:
Add rollback script, move auto-changelog script (#6252)

whymarrh committed:
Bump gaba version to avoid broken eth-contract-metadata

whymarrh committed:
Merge pull request #6327 from MetaMask/eth-contract-metadata

Update eth-contract-metadata dependency and sub-dependencies
